src/erlfmt.erl

```diff
@@ -78,39 +82,71 @@ format_file(FileName, {Pragma, Out}) ->
         {error, Error} -> {error, Error}
     end.
 
--spec format_string(string(), integer()) -> string().
-format_string(String, PageWidth) ->
+-spec format_string(string(), integer(), pragma()) -> string().
+format_string(String, PageWidth, Pragma) ->
     {ok, Nodes, []} = read_nodes_string("nofile", String),
-    [$\n | Formatted] = format_nodes(Nodes, PageWidth),
+    NodesWithPragma =
+        case Pragma of
+            insert -> insert_pragma_nodes(Nodes);
+            _ -> Nodes
+        end,
+    [$\n | Formatted] = format_nodes(NodesWithPragma, PageWidth),
     unicode:characters_to_list(Formatted).
 
--spec contains_pragma_file(file:name_all()) -> boolean().
-contains_pragma_file(FileName) ->
-    read_file(FileName, fun (File) ->
-        read_pragma_nodes(erlfmt_scan:io_node(File), FileName, [], [])
-    end).
-
--spec contains_pragma_string(file:name_all(), string()) -> boolean().
+-spec contains_pragma_string(file:name_all(), string()) -> boolean() | {error, error_info()}.
 contains_pragma_string(FileName, String) ->
-    read_pragma_nodes(erlfmt_scan:string_node(String), FileName, [], []).
+    try read_nodes(erlfmt_scan:string_node(String), FileName, require) of
+        {skip, _} -> false;
+        _ -> true
+    catch
+        {error, Error} -> {error, Error}
+    end.
 
-read_pragma_nodes({ok, Tokens, Comments, Cont}, FileName, _Acc, Warnings0) ->
-    {Node, _Warnings} = parse_nodes(Tokens, Comments, FileName, Cont, Warnings0),
-    contains_pragma_node(Node);
-read_pragma_nodes(_, _, _, _) ->
-    false.
+is_shebang_node({raw_string, _, "#!" ++ _}) -> true;
+is_shebang_node(_) -> false.
```

Comment:
We're converting a special `shebang` node into raw string in here https://github.com/WhatsApp/erlfmt/blob/1b3a001e1f64616a4b484ef4a15bbf681e81062d/src/erlfmt.erl#L255-L256. Do you think it would make sense to keep the special node around longer to simplify this function or even remove it and pattern match on the `shebang` node directly? We would need to change `format_node` to handle the new `shebang` in the same way that it's handling `raw_string` right now

Reply:
That sounds great.  How about I tackle that in a follow up commit?

Reply:
On second thought, I saw it was easier than expected.  So I have now added it in this commit.